File: dan/Problems/Medium/Graph/207. Course Schedule/solution.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Stack<Integer> stack = new Stack<Integer>();
        int[][] graph = new int[numCourses][numCourses];
        int[] indegree = new int[numCourses];
        int cnt = 0;
        for(int i = 0; i < prerequisites.length; i++){
            int pre = prerequisites[i][1];
            int lat = prerequisites[i][0];
            graph[pre][lat] = 1;
            indegree[lat]++;
        }
        for(int i = 0; i < numCourses; i++){
            if(indegree[i] == 0){
                stack.push(i);
            }
        }
        while(!stack.isEmpty()){
            int course = stack.pop();
            cnt++;
            for(int i = 0; i < numCourses; i++){
                if(graph[course][i] == 1){
                    indegree[i]--;
                    if(indegree[i] == 0){
                        stack.push(i);
                    }
                }
            }
        }
        return cnt == numCourses;
    }
}

File: dan/Problems/Medium/Graph/207. Course Schedule/solution.py
class Solution(object):
    #DFS. Using stack.
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        stack, arc_num, cnt = [], len(prerequisites), 0
        
        #Initialize the graph
        graph = [[] for i in range(numCourses)]
        indegree = [0 for i in range(numCourses)]
        
        #Add edges to the graph
        for pair in prerequisites:
            pre, lat = pair[1], pair[0]
            graph[pre].append(lat)
            indegree[lat] += 1
            
        #Initialize the stack
        for idx, ind in enumerate(indegree):
            if ind == 0:
                stack.append(idx)
                
        while stack:
            course = stack.pop()
            cnt += 1
            for lat in graph[course]:
                indegree[lat] -= 1
                if indegree[lat] == 0:
                    stack.append(lat)
                    
        return cnt == numCourses
            
        
        
